File: app/src/main/java/net/yuntara/dspplayer/StreamPlayer.java
package net.yuntara.dspplayer;


import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Timer;
import java.util.TimerTask;
import java.util.stream.Stream;

import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;
import android.media.MediaCodec;
import android.media.MediaExtractor;
import android.media.MediaFormat;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Message;
import android.support.v4.media.session.PlaybackStateCompat;

import android.util.Log;
import android.content.Context;
import android.net.Uri;


/**
 * Created by yuntara on 2016/11/03.
 */

public class StreamPlayer {
    protected MediaExtractor extractor;
    protected MediaCodec codec;
    protected AudioTrack audioTrack;
    private boolean doStop;
    private boolean paused = false;
    private Context songContext;
    private Uri songUri;
    protected int inputBufIndex;
    protected int bufIndexCheck;
    protected int lastInputBufIndex;
    private boolean finished = true;
    private final int SIZEX = 65536;
    private boolean stopped = false;
    private Context mainContext;
    private MusicService msc;
    private float datachunk[];
    private float x[],y[];

    private short before_dataL[],
            before_dataR[];
    private short audio_dataL[],
            audio_dataR[];
    private int audio_pos = 0;
    private float seekTo = -1;
    public enum State {
        Retrieving, // retrieving music (filling buffer)
        Stopped,    // player is stopped and not prepared to play
        Playing,    // playback active
    };
    private State mState = State.Stopped;
    private String LOG_TAG = "StreamPlayer";


    public void setDataSource(Context context,Uri uri){
        songContext = context;
        songUri = uri;

    }
    public StreamPlayer(MusicService m){
        //native_lib initialize func
        cfftinit(SIZEX*2);
        msc = m;
    }
    public void setPosition(float seekTo){
        this.seekTo = seekTo;
    }
    public void loadFilter(Context context){
        ///*
        float filR1_r[];
        float filR2_r[];
        float filL1_r[];
        float filL2_r[];
        float filR1_i[];
        float filR2_i[];
        float filL1_i[];
        float filL2_i[];
        filR1_r = new float[SIZEX+1];
        filR1_i = new float[SIZEX+1];
        filR2_r = new float[SIZEX+1];
        filR2_i = new float[SIZEX+1];
        filL1_r = new float[SIZEX+1];
        filL1_i = new float[SIZEX+1];
        filL2_r = new float[SIZEX+1];
        filL2_i = new float[SIZEX+1];

        before_dataL = new short[SIZEX];
        before_dataR = new short[SIZEX];

        audio_dataL = new short[SIZEX];
        audio_dataR = new short[SIZEX];

        datachunk = new float[SIZEX*2];
        x = new float[SIZEX*2];
        y = new float[SIZEX*2];

        final int id = context.getResources().getIdentifier("out", "raw", context.getPackageName());
        if (id == 0) {    //エラーにはならない
            return;
        }
        InputStream is = context.getResources().openRawResource(id);
        DataInputStream ds = new DataInputStream(is);
        try {

            for (int i = 0; i <= SIZEX; i++) {
                filR1_r[i] = (float)ds.readDouble();
                filR1_i[i] = (float)ds.readDouble();
            }
            for (int i = 0; i <= SIZEX; i++) {
                filR2_r[i] = (float)ds.readDouble();
                filR2_i[i] = (float)ds.readDouble();
            }
            for (int i = 0; i <= SIZEX; i++) {
                filL1_r[i] = (float)ds.readDouble();
                filL1_i[i] = (float)ds.readDouble();
            }
            for (int i = 0; i <= SIZEX; i++) {
                filL2_r[i] = (float)ds.readDouble();
                filL2_i[i] = (float)ds.readDouble();
            }
            csetfil(filR1_r,filR1_i,filR2_r,filR2_i,filL1_r,filL1_i,filL2_r,filL2_i);
        }catch(IOException e){

            Arrays.fill(filR1_r,0);
            Arrays.fill(filR1_i,0);
            Arrays.fill(filL1_r,0);
            Arrays.fill(filL1_i,0);
            Arrays.fill(filR2_r,0);
            Arrays.fill(filR2_i,0);
            Arrays.fill(filL2_r,0);
            Arrays.fill(filL2_i,0);
            csetfil(filR1_r,filR1_i,filR2_r,filR2_i,filL1_r,filL1_i,filL2_r,filL2_i);
        }

    }
    //private void queueChunk(short[] chunk,int csize) {
    private void queueChunk(java.nio.ShortBuffer shortBuffer,int csize) {
        //csize = (datalength) * (2channel)
        //畳みこみできるサイズになるまでバッファに貯める
        boolean process = false;

        int samples = csize / 2;

        //変数名のLRが間違ってますが、フィルタ作成の方のプログラムから間違っているので、そのうちrefactorします
        for (int i = 0; i < samples; i++) {
            //audio_dataL[audio_pos + i] = chunk[2 * (i) + 1];//右音声
            //audio_dataR[audio_pos + i] = chunk[2 * (i)];    //左音声

            audio_dataR[audio_pos + i] = shortBuffer.get();    //左音声
            audio_dataL[audio_pos + i] = shortBuffer.get();   //右音声
            if (audio_pos + i >= SIZEX - 1) {
                audio_pos = i;
                process = true;
                break;
            }
        }
        //まだ足りない
        if (!process) {
            audio_pos += samples;
            return;
        }

        //overlap-save法のために、2ブロック分のデータが必要　　(overlap-addだと畳みこみ後に連結処理が入るので面倒くさいです)
        for (int j = 0; j < SIZEX * 2; j++) {
            //前半ブロックは前回のデータ
            if (j<SIZEX) {
                x[j] = before_dataL[j];
                y[j] = before_dataR[j];
            }
            //後半ブロックが今回のデータ
            else {
                x[j] = audio_dataL[(j - SIZEX)];//
                y[j] = audio_dataR[(j - SIZEX)];
            }
        }

        //次回の畳みこみに使う分をコピーしておく
        System.arraycopy(audio_dataL, 0, before_dataL, 0, SIZEX);
        System.arraycopy(audio_dataR, 0, before_dataR, 0, SIZEX);

        //native_lib.cpp で畳みこみします
        ccomboluteRL(x,y);

        //直線畳みこみ（ifftの前半部分）しか使わない
        for (int j = 0; j < SIZEX; j++) {
            datachunk[2*j]   = x[j];
            datachunk[2*j+1] = y[j];
        }

        /*
          AudioTrack.WRITE_BLOCKING だとバッファーが書き込める状態になるまでブロックされるので、特別に処理をしなくても連続で再生される。
          ただし、書き込むサイズとAudioTrackのバッファーサイズ違うと面倒くさいので合わせておく。
        */
        audioTrack.write(datachunk,0,2*SIZEX,AudioTrack.WRITE_BLOCKING);

        //次回の書き込むオフセット
        audio_pos = samples - audio_pos - 1;

        //今回使わなかった分をバッファーの先頭にとっておきます。
        for (int i = 0; i < audio_pos; i++) {
            audio_dataR[i] = shortBuffer.get();
            audio_dataL[i] = shortBuffer.get();
        }



    }
    public void pause(){
        paused = !paused;
    }
    public void play()
    {

        mState = State.Retrieving;
        //mDelegateHandler.onRadioPlayerBuffering(MP3RadioStreamPlayer.this);

        doStop = false;
        finished = false;
        bufIndexCheck = 0;

        lastInputBufIndex = -1;
        DecodeOperation dcd = new DecodeOperation();
        dcd.setOnCallBack(new CallBackTask(){

            @Override
            public void CallBack() {

                super.CallBack();


                msc.OnPlayEnd();
                stopped = false;


            }

        });
        dcd.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);

    }
    private void decodeLoop(DecodeOperation parent) throws  IOException
    {


        // extractor gets information about the stream
        extractor = new MediaExtractor();
        try {
            extractor.setDataSource(songContext,songUri,null);
        } catch (Exception e) {

            return;
        }

        MediaFormat format = extractor.getTrackFormat(0);
        String mime = format.getString(MediaFormat.KEY_MIME);

        // the actual decoder
        codec = MediaCodec.createDecoderByType(mime);
        codec.configure(format, null /* surface */, null /* crypto */, 0 /* flags */);
        codec.start();

        int sampleRate = format.getInteger(MediaFormat.KEY_SAMPLE_RATE);

        //int bits = format.getInteger(MediaFormat.KEY_PCM_ENCODING);

        if(sampleRate != 44100 ){
            this.mState = State.Stopped;
            return;
        }


        //int maxSize = format.getInteger(MediaFormat.KEY_MAX_INPUT_SIZE);


        // create our AudioTrack instance
        if(audioTrack == null) {

            audioTrack = new AudioTrack(
                    AudioManager.STREAM_MUSIC,
                    sampleRate,
                    AudioFormat.CHANNEL_OUT_STEREO,
                    AudioFormat.ENCODING_PCM_FLOAT,
                /*AudioTrack.getMinBufferSize (
                        sampleRate,
                        AudioFormat.CHANNEL_OUT_STEREO,
                        AudioFormat.ENCODING_PCM_16BIT
                )
                */
                    //2channel * bufferlength * sizeof(float)
                    2 * SIZEX * 4
                    ,
                    AudioTrack.MODE_STREAM
            );
        }

        // start playing, we will feed you later
        audioTrack.play();
        extractor.selectTrack(0);

        // start decoding
        final long kTimeOutUs = 10000;
        MediaCodec.BufferInfo info = new MediaCodec.BufferInfo();
        boolean sawInputEOS = false;
        boolean sawOutputEOS = false;
        int noOutputCounter = 0;
        int noOutputCounterLimit = 50;
        long duration = format.getLong(MediaFormat.KEY_DURATION);

        while (!sawOutputEOS && noOutputCounter < noOutputCounterLimit && !doStop) {
            while(paused && !doStop){
                try {
                    Thread.sleep(100);
                    //info.


                }catch(InterruptedException e){

                }
            }

            if(this.seekTo>0){
                long pos = (long) (duration * this.seekTo/1.0f);
                Log.i(LOG_TAG,"seek to "+pos +" = "+duration +"*"+seekTo+"/1000.0f" );
                this.seekTo=-1;
                if(pos<duration) {

                    //extractor.
                    //extractor.seekTo(pos, MediaExtractor.SEEK_TO_PREVIOUS_SYNC);
                    long starttime = System.currentTimeMillis();

                    long presentationTimeUs = extractor.getSampleTime();
                    while(presentationTimeUs < pos && System.currentTimeMillis() < starttime + 3000) {
                        extractor.advance();
                        presentationTimeUs = extractor.getSampleTime();
                    }
                    //extractor.
/*
I/StreamPlayer: seek to 7960 = 274506666*0.029/1000.0f
E/FileSource: seek to 12238042466020 failed
E/FileSource: seek to 12238042458078 failed
 */
                }

            }

            //Log.i(LOG_TAG, "loop ");
            noOutputCounter++;
            if (!sawInputEOS) {

                inputBufIndex = codec.dequeueInputBuffer(kTimeOutUs);
                bufIndexCheck++;
                // Log.d(LOG_TAG, " bufIndexCheck " + bufIndexCheck);
                if (inputBufIndex >= 0) {


                    ByteBuffer dstBuf = codec.getInputBuffer(inputBufIndex);

                    int sampleSize =
                            extractor.readSampleData(dstBuf, 0 /* offset */);
                    //extractor.
                    long presentationTimeUs = 0;

                    if (sampleSize < 0) {
                        Log.d(LOG_TAG, "saw input EOS.");
                        sawInputEOS = true;
                        sampleSize = 0;
                    } else {
                        presentationTimeUs = extractor.getSampleTime();
                    }
                    // can throw illegal state exception (???)

                    //Progress
                    float prog =  ((float)presentationTimeUs / duration);
                    parent.progress(prog);

                    codec.queueInputBuffer(
                            inputBufIndex,
                            0 /* offset */,
                            sampleSize,
                            presentationTimeUs,
                            sawInputEOS ? MediaCodec.BUFFER_FLAG_END_OF_STREAM : 0);



                    if (!sawInputEOS) {
                        extractor.advance();
                    }
                }
                else
                {

                }
            }

            int res = codec.dequeueOutputBuffer(info, kTimeOutUs);

            if (res >= 0) {
                if (info.size > 0) {
                    noOutputCounter = 0;
                }
                int outputBufIndex = res;

                ByteBuffer buf = codec.getOutputBuffer(outputBufIndex);
                if(info.size > 0){
                    queueChunk(buf.asShortBuffer(),info.size/2);
                    this.mState = State.Playing;
                }
                buf.clear();
                codec.releaseOutputBuffer(outputBufIndex, false /* render */);
                if ((info.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {

                    sawOutputEOS = true;
                }
            } else if (res == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                MediaFormat oformat = codec.getOutputFormat();

            } else {

            }
        }
        if(doStop){
            parent.setUseCallBack(false);
        }
        this.mState = State.Stopped;
    }
    public void reset(){
        stop();
    }
    public void release()
    {
        stop();
        relaxResources(false);
    }

    private void relaxResources(Boolean release)
    {
        if(codec != null)
        {
            if(release)
            {
                codec.stop();
                codec.release();
                codec = null;
            }

        }
        if(audioTrack != null)
        {
            audioTrack.flush();
            audioTrack.release();
            audioTrack = null;
        }
    }
    public void stop()
    {
        doStop = true;

        if(mState != State.Stopped){
            stopped = true;
        }
        while(mState != State.Stopped) {
            try{
                Thread.sleep(10);
            }catch(InterruptedException e){
                break;
            }

        }

    }
    public static class CallBackTask {
        public void CallBack() {
        }
    }
    private class DecodeOperation extends AsyncTask<Void, Float, Void> {
        private CallBackTask callbacktask;
        private boolean useCallBack = true;
        @Override
        protected Void doInBackground(Void... values) {
            try {
                StreamPlayer.this.decodeLoop(this);
            } catch (IOException e) {
                //TODO catch IOException at decode
            }
            finished=true;
            return null;
        }
        public void progress(float prg){
            onProgressUpdate(prg);
        }
        @Override
        protected void onPostExecute(Void result)  {
            super.onPostExecute(null);
            if(useCallBack) {
                callbacktask.CallBack();
            }
        }
        public boolean getUseCallBack(){
            return useCallBack;
        }
        public void setUseCallBack(boolean useCallBack){
            this.useCallBack = useCallBack;
        }
        public void setOnCallBack(CallBackTask _cbj) {
            callbacktask = _cbj;
        }
        @Override
        protected void onPreExecute() {
        }

        @Override
        protected void onProgressUpdate(Float... values) {
            StreamPlayer.this.msc.onProgress(values[0]);

        }
    }
    public native void cfftinit(int size);
    //public native void cfft(float x[],float y[],boolean isReverse);
    //public native void cifft(float x[],float y[]);
    public native void csetfil(float r1r[],float r1i[],float r2r[],float r2i[],
                                  float l1r[],float l1i[],float l2r[],float l2i[]
                               );
    public native void ccomboluteRL(float x[],float y[]);
    //public native void cadjust(float x[],float y[]);

    static {
        System.loadLibrary("native-lib");
    }
}

File: app/src/main/java/net/yuntara/dspplayer/StreamPlayer.kt
package net.yuntara.dspplayer


import java.io.DataInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.util.Arrays
import java.util.Timer
import java.util.TimerTask
import java.util.stream.Stream

import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.media.MediaCodec
import android.media.MediaExtractor
import android.media.MediaFormat
import android.os.AsyncTask
import android.os.Handler
import android.os.Message
import android.support.v4.media.session.PlaybackStateCompat

import android.util.Log
import android.content.Context
import android.net.Uri


/**
 * Created by yuntara on 2016/11/03.
 */

class StreamPlayer(private val msc: MusicService) {
    protected var extractor: MediaExtractor? = null
    protected var codec: MediaCodec? = null
    protected var audioTrack: AudioTrack? = null
    private var doStop: Boolean = false
    private var paused = false
    private var songContext: Context? = null
    private var songUri: Uri? = null
    protected var inputBufIndex: Int = 0
    protected var bufIndexCheck: Int = 0
    protected var lastInputBufIndex: Int = 0
    private var finished = true
    private val SIZEX = 65536/2
    private var stopped = false
    private val mainContext: Context? = null
    private var datachunk: FloatArray? = null
    private var x: FloatArray? = null
    private var y: FloatArray? = null
    private var filterName = "outthree"
    private var before_dataL: ShortArray? = null
    private var before_dataR: ShortArray? = null
    private var audio_dataL: ShortArray? = null
    private var audio_dataR: ShortArray? = null
    private var audio_pos = 0
    private var seekTo = -1f

    enum class State {
        Retrieving, // retrieving music (filling buffer)
        Stopped, // player is stopped and not prepared to play
        Playing
        // playback active
    }

    private var mState = State.Stopped
    private val LOG_TAG = "StreamPlayer"


    fun setDataSource(context: Context, uri: Uri) {
        songContext = context
        songUri = uri

    }

    fun setFilter(context: Context, filname: String) {
        paused = true
        filterName = filname
        loadFilter(context)
        paused = false
    }

    init {
        //native_lib initialize func
        cfftinit(SIZEX * 2)
    }

    fun setPosition(seekTo: Float) {
        this.seekTo = seekTo
    }

    fun loadFilter(context: Context) {
        ///*
        val filR1_r: FloatArray
        val filR2_r: FloatArray
        val filL1_r: FloatArray
        val filL2_r: FloatArray
        val filR1_i: FloatArray
        val filR2_i: FloatArray
        val filL1_i: FloatArray
        val filL2_i: FloatArray
        filR1_r = FloatArray(SIZEX + 1)
        filR1_i = FloatArray(SIZEX + 1)
        filR2_r = FloatArray(SIZEX + 1)
        filR2_i = FloatArray(SIZEX + 1)
        filL1_r = FloatArray(SIZEX + 1)
        filL1_i = FloatArray(SIZEX + 1)
        filL2_r = FloatArray(SIZEX + 1)
        filL2_i = FloatArray(SIZEX + 1)

        before_dataL = ShortArray(SIZEX)
        before_dataR = ShortArray(SIZEX)

        audio_dataL = ShortArray(SIZEX)
        audio_dataR = ShortArray(SIZEX)

        datachunk = FloatArray(SIZEX * 2)
        x = FloatArray(SIZEX * 2)
        y = FloatArray(SIZEX * 2)

        val id = context.resources.getIdentifier(filterName, "raw", context.packageName)
        if (id == 0) {    //エラーにはならない
            return
        }
        val `is` = context.resources.openRawResource(id)
        val ds = DataInputStream(`is`)
        try {

            for (i in 0..SIZEX) {
                filR1_r[i] = ds.readDouble().toFloat()
                filR1_i[i] = ds.readDouble().toFloat()
            }
            for (i in 0..SIZEX) {
                filR2_r[i] = ds.readDouble().toFloat()
                filR2_i[i] = ds.readDouble().toFloat()
            }
            for (i in 0..SIZEX) {
                filL1_r[i] = ds.readDouble().toFloat()
                filL1_i[i] = ds.readDouble().toFloat()
            }
            for (i in 0..SIZEX) {
                filL2_r[i] = ds.readDouble().toFloat()
                filL2_i[i] = ds.readDouble().toFloat()
            }
            csetfil(filR1_r, filR1_i, filR2_r, filR2_i, filL1_r, filL1_i, filL2_r, filL2_i)
        } catch (e: IOException) {

            Arrays.fill(filR1_r, 0f)
            Arrays.fill(filR1_i, 0f)
            Arrays.fill(filL1_r, 0f)
            Arrays.fill(filL1_i, 0f)
            Arrays.fill(filR2_r, 0f)
            Arrays.fill(filR2_i, 0f)
            Arrays.fill(filL2_r, 0f)
            Arrays.fill(filL2_i, 0f)
            csetfil(filR1_r, filR1_i, filR2_r, filR2_i, filL1_r, filL1_i, filL2_r, filL2_i)
        }

    }

    //private void queueChunk(short[] chunk,int csize) {
    private fun queueChunk(shortBuffer: java.nio.ShortBuffer, csize: Int) {
        //csize = (datalength) * (2channel)
        //畳みこみできるサイズになるまでバッファに貯める
        var process = false

        val samples = csize / 2

        //変数名のLRが間違ってますが、フィルタ作成の方のプログラムから間違っているので、そのうちrefactorします
        for (i in 0..samples - 1) {
            //audio_dataL[audio_pos + i] = chunk[2 * (i) + 1];//右音声
            //audio_dataR[audio_pos + i] = chunk[2 * (i)];    //左音声

            audio_dataR!![audio_pos + i] = shortBuffer.get()    //左音声
            audio_dataL!![audio_pos + i] = shortBuffer.get()   //右音声
            if (audio_pos + i >= SIZEX - 1) {
                audio_pos = i
                process = true
                break
            }
        }
        //まだ足りない
        if (!process) {
            audio_pos += samples
            return
        }

        //overlap-save法のために、2ブロック分のデータが必要　　(overlap-addだと畳みこみ後に連結処理が入るので面倒くさいです)
        for (j in 0..SIZEX * 2 - 1) {
            //前半ブロックは前回のデータ
            if (j < SIZEX) {
                x!![j] = before_dataL!![j].toFloat()
                y!![j] = before_dataR!![j].toFloat()
            } else {
                x!![j] = audio_dataL!![j - SIZEX].toFloat()//
                y!![j] = audio_dataR!![j - SIZEX].toFloat()
            }//後半ブロックが今回のデータ
        }

        //次回の畳みこみに使う分をコピーしておく
        System.arraycopy(audio_dataL!!, 0, before_dataL!!, 0, SIZEX)
        System.arraycopy(audio_dataR!!, 0, before_dataR!!, 0, SIZEX)

        //native_lib.cpp で畳みこみします
        ccomboluteRL(x!!, y!!)

        //直線畳みこみ（ifftの前半部分）しか使わない
        for (j in 0..SIZEX - 1) {
            datachunk!![2 * j] = x!![j]
            datachunk!![2 * j + 1] = y!![j]
        }

        /*
          AudioTrack.WRITE_BLOCKING だとバッファーが書き込める状態になるまでブロックされるので、特別に処理をしなくても連続で再生される。
          ただし、書き込むサイズとAudioTrackのバッファーサイズ違うと面倒くさいので合わせておく。
        */
        audioTrack!!.write(datachunk!!, 0, 2 * SIZEX, AudioTrack.WRITE_BLOCKING)

        //次回の書き込むオフセット
        audio_pos = samples - audio_pos - 1

        //今回使わなかった分をバッファーの先頭にとっておきます。
        for (i in 0..audio_pos - 1) {
            audio_dataR!![i] = shortBuffer.get()
            audio_dataL!![i] = shortBuffer.get()
        }


    }

    fun pause() {
        paused = !paused
    }

    fun play() {

        mState = State.Retrieving
        //mDelegateHandler.onRadioPlayerBuffering(MP3RadioStreamPlayer.this);

        doStop = false
        finished = false
        bufIndexCheck = 0

        lastInputBufIndex = -1

        val dcd = DecodeOperation()
        dcd.setOnCallBack(object : CallBackTask() {

            override fun CallBack() {

                super.CallBack()


                msc.OnPlayEnd()
                stopped = false


            }

        })
        dcd.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)

    }

    @Throws(IOException::class)
    private fun decodeLoop(parent: DecodeOperation) {


        // extractor gets information about the stream
        extractor = MediaExtractor()
        try {
            extractor!!.setDataSource(songContext!!, songUri!!, null)
        } catch (e: Exception) {

            return
        }
        val format = extractor!!.getTrackFormat(0)
        val mime = format.getString(MediaFormat.KEY_MIME)

        // the actual decoder
        codec = MediaCodec.createDecoderByType(mime)
        codec!!.configure(format, null, null, 0 /* flags */)/* surface *//* crypto */
        codec!!.start()

        val sampleRate = format.getInteger(MediaFormat.KEY_SAMPLE_RATE)

        //int bits = format.getInteger(MediaFormat.KEY_PCM_ENCODING);

        if (sampleRate != 44100) {
            this.mState = State.Stopped
            return
        }


        //int maxSize = format.getInteger(MediaFormat.KEY_MAX_INPUT_SIZE);


        // create our AudioTrack instance
        if (audioTrack == null) {

            audioTrack = AudioTrack(
                    AudioManager.STREAM_MUSIC,
                    sampleRate,
                    AudioFormat.CHANNEL_OUT_STEREO,
                    AudioFormat.ENCODING_PCM_FLOAT,
                    /*AudioTrack.getMinBufferSize (
                        sampleRate,
                        AudioFormat.CHANNEL_OUT_STEREO,
                        AudioFormat.ENCODING_PCM_16BIT
                )
                */
                    //2channel * bufferlength * sizeof(float)
                    2 * SIZEX * 4,
                    AudioTrack.MODE_STREAM
            )
        }

        // start playing, we will feed you later
        audioTrack!!.play()
        extractor!!.selectTrack(0)

        // start decoding
        val kTimeOutUs: Long = 10000
        val info = MediaCodec.BufferInfo()
        var sawInputEOS = false
        var sawOutputEOS = false
        var noOutputCounter = 0
        val noOutputCounterLimit = 50
        val duration = format.getLong(MediaFormat.KEY_DURATION)

        while (!sawOutputEOS && noOutputCounter < noOutputCounterLimit && !doStop) {
            while (paused && !doStop) {
                try {
                    Thread.sleep(100)
                    //info.


                } catch (e: InterruptedException) {

                }

            }

            if (this.seekTo > 0) {
                val pos = (duration * this.seekTo / 1.0f).toLong()
                Log.i(LOG_TAG, "seek to $pos = $duration*$seekTo/1000.0f")
                this.seekTo = -1f
                if (pos < duration) {

                    //extractor.
                    //extractor.seekTo(pos, MediaExtractor.SEEK_TO_PREVIOUS_SYNC);
                    val starttime = System.currentTimeMillis()

                    var presentationTimeUs = extractor!!.sampleTime
                    while (presentationTimeUs < pos && System.currentTimeMillis() < starttime + 3000) {
                        extractor!!.advance()
                        presentationTimeUs = extractor!!.sampleTime
                    }
                    //extractor.
                    /*
I/StreamPlayer: seek to 7960 = 274506666*0.029/1000.0f
E/FileSource: seek to 12238042466020 failed
E/FileSource: seek to 12238042458078 failed
 */
                }

            }

            //Log.i(LOG_TAG, "loop ");
            noOutputCounter++
            if (!sawInputEOS) {

                inputBufIndex = codec!!.dequeueInputBuffer(kTimeOutUs)
                bufIndexCheck++
                // Log.d(LOG_TAG, " bufIndexCheck " + bufIndexCheck);
                if (inputBufIndex >= 0) {


                    val dstBuf = codec!!.getInputBuffer(inputBufIndex)

                    var sampleSize = extractor!!.readSampleData(dstBuf!!, 0 /* offset */)
                    //extractor.
                    var presentationTimeUs: Long = 0

                    if (sampleSize < 0) {
                        Log.d(LOG_TAG, "saw input EOS.")
                        sawInputEOS = true
                        sampleSize = 0
                    } else {
                        presentationTimeUs = extractor!!.sampleTime
                    }
                    // can throw illegal state exception (???)

                    //Progress
                    val prog = presentationTimeUs.toFloat() / duration
                    parent.progress(prog)

                    codec!!.queueInputBuffer(
                            inputBufIndex,
                            0 /* offset */,
                            sampleSize,
                            presentationTimeUs,
                            if (sawInputEOS) MediaCodec.BUFFER_FLAG_END_OF_STREAM else 0)



                    if (!sawInputEOS) {
                        extractor!!.advance()
                    }
                } else {

                }
            }

            val res = codec!!.dequeueOutputBuffer(info, kTimeOutUs)

            if (res >= 0) {
                if (info.size > 0) {
                    noOutputCounter = 0
                }
                val outputBufIndex = res

                val buf = codec!!.getOutputBuffer(outputBufIndex)
                if (info.size > 0) {
                    queueChunk(buf!!.asShortBuffer(), info.size / 2)
                    this.mState = State.Playing
                }
                buf!!.clear()
                codec!!.releaseOutputBuffer(outputBufIndex, false /* render */)
                if (info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) {

                    sawOutputEOS = true
                }
            } else if (res == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                val oformat = codec!!.outputFormat

            } else {

            }
        }
        if (doStop) {
            parent.useCallBack = false
        }
        this.mState = State.Stopped
    }

    fun reset() {
        stop()
    }

    fun release() {
        stop()
        relaxResources(false)
    }

    private fun relaxResources(release: Boolean) {
        if (codec != null) {
            if (release) {
                codec!!.stop()
                codec!!.release()
                codec = null
            }

        }
        if (audioTrack != null) {
            audioTrack!!.flush()
            audioTrack!!.release()
            audioTrack = null
        }
    }

    fun stop() {
        doStop = true

        if (mState != State.Stopped) {
            stopped = true
        }
        while (mState != State.Stopped) {
            try {
                Thread.sleep(10)
            } catch (e: InterruptedException) {
                break
            }

        }

    }

    open class CallBackTask {
        open fun CallBack() {}
    }

    private inner class DecodeOperation : AsyncTask<Void, Float, Void>() {
        private var callbacktask: CallBackTask? = null
        var useCallBack = true
        override fun doInBackground(vararg values: Void): Void? {
            try {
                this@StreamPlayer.decodeLoop(this)
            } catch (e: IOException) {
                //TODO catch IOException at decode
            }

            finished = true
            return null
        }

        fun progress(prg: Float) {
            onProgressUpdate(prg)
        }

        override fun onPostExecute(result: Void?) {
            super.onPostExecute(result)
            if (useCallBack) {
                callbacktask?.CallBack()
            }
        }

        fun setOnCallBack(_cbj: CallBackTask) {
            callbacktask = _cbj
        }

        override fun onPreExecute() {}

        protected fun onProgressUpdate(vararg values: Float) {
            this@StreamPlayer.msc.onProgress(values[0])

        }
    }

    external fun cfftinit(size: Int)
    //public native void cfft(float x[],float y[],boolean isReverse);
    //public native void cifft(float x[],float y[]);
    external fun csetfil(r1r: FloatArray, r1i: FloatArray, r2r: FloatArray, r2i: FloatArray,
                         l1r: FloatArray, l1i: FloatArray, l2r: FloatArray, l2i: FloatArray
    )

    external fun ccomboluteRL(x: FloatArray, y: FloatArray)

    companion object {
        //public native void cadjust(float x[],float y[]);

        init {
            System.loadLibrary("native-lib")
        }
    }
}
